mozilla-mobile/firefox-android
# Bug 1867539 - Add UI Test Tooling Activation

Overview

This PR introduces a significant update to our testing pipeline by automating test execution and integrating notification mechanisms. It focuses on enhancing our TestRail interaction and Slack notifications to streamline our testing workflow.

Key Updates

1. Test Automation Execution
- When a new release branch is created or pushed to, the Taskcluster task will run.
- The Firebase Tests will run, reporting a `PASS` or `FAIL` for use with the TestRail Connection Library.
- Environment Variables will be created for the `PRODUCT_TYPE`, `RELEASE_TYPE`, and `TEST_STATUS`.
- The TestRail Connection Library then establishes a connection with the Mozilla TestRail instance using authenticated credentials.
- Creates new Milestones and Test Runs in TestRail for tracking automated test results.
- Posts test results to the corresponding TestRail Milestone as a Test Run.
- Updates test cases based on automated smoke test results for different product releases.

2. Notification System:

- Implements Slack notifications for test status, ensuring immediate feedback on test results.
- On Test Failure, posts Slack message with debugging information for triage.
- On Sucess, posts Slack message with Milestone and Test Run reporting results.

Phased Implementation Strategy

We're adopting a phased approach to monitor changes and ensure seamless integration:

1. MVP Phase:

- Trigger automated test execution.
- Post data to a dummy Testrail project (FirefoxTV).
- Send notifications to a sandbox Slack channel.
- Goals: Validate triggering mechanisms, frequency, and Slack notifications.

2. Production Project Integration:

- Direct automation to the production project: Fenix.
- Continue notifications to the sandbox channel.

3. Final Phase:

- Parameterize all project variables.
- Include Focus Android in the testing suite.
- Switch notifications to the #mobiletesteng Slack channel.

Fallback Strategy

A fallback strategy is included to turn off the workflow/script in case of unforeseen issues, ensuring system stability.

Updates in kind.yml File

The kind.yml file has been updated to reflect these changes:

- New environment variables have been added to the 'Focus' and 'Firefox' beta release tasks.
- The Python script call has been integrated to write to TestRail when Firebase jobs pass.

Purpose

The primary goal of this PR is to streamline our testing process, reduce manual intervention, and improve our response time to test results. This will also reduce the manual work required during the release process for our reporting and observability.

### Pull Request checklist

- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

















### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1867539
